torch/_library/abstract_impl_class.py

```diff
@@ -0,0 +1,153 @@
+
+    It's in the same spirit of registering an abstract implementation for
+    an operator with impl_abstract but with the difference that it
+    associates a fake class with the original torch bind class (registered
+    with torch::class_).  In this way, , object of the class can be properly
+    guarded and tracked by components in PT2 stack such as Dynamo and AOTAutograd.
+
+    This API may be used as a decorator (see examples).
+
+    Examples:
+        # For a torch Bind class Foo defined in test_custom_class_registration.cpp:
+        TORCH_LIBRARY(_TorchScriptTesting, m) {
+            m.class_<Foo>("_Foo")
+                .def(torch::init<int64_t, int64_t>())
+                // .def(torch::init<>())
+                .def("info", &Foo::info)
+                .def("increment", &Foo::increment)
+                .def("add", &Foo::add)
+                .def("add_tensor", &Foo::add_tensor)
+                .def("__eq__", &Foo::eq)
+                .def("combine", &Foo::combine)
+                .def_pickle(
+                    [](c10::intrusive_ptr<Foo> self) { // __getstate__
+                      return std::vector<int64_t>{self->x, self->y};
+                    },
+                    [](std::vector<int64_t> state) { // __setstate__
+                      return c10::make_intrusive<Foo>(state[0], state[1]);
+                    })
+                .def_meta([](c10::intrusive_ptr<Foo> self) { // __get_metadata__
+                  return std::vector<int64_t>{self->x, self->y};
+                });
+        # We could register a fake class FakeFoo in Python as follows:
+        import torch
+
+        @torch._library.impl_abstract_class("_TorchScriptTesting::_Foo")
+        class FakeFoo:
+            def __init__(self, x, y):
+                self.x = x
+                self.y = y
+
+            @classmethod
+            def from_real(cls, foo_obj):
+                # __get_metadata__ is defined by .def_meta in C++
+                x, y = foo_obj.__get_metadata__()
+                return cls(x, y)
+
+            def add_tensor(self, z):
+                return (self.x + self.y) * z1
+
+    Temporal Limitations:
+        - We don't support method call on the script object yet. Please use custom op to manipulate them. Please
+          implement a custom op that takes the script object as input and call the method in the custom op.
+
+    Examples:
+        # CPU impl in test_custom_class_registration.cpp:
+
+        at::Tensor takes_foo(c10::intrusive_ptr<Foo> foo, at::Tensor x) {
+          return foo->add_tensor(x);
+        }
+        TORCH_LIBRARY_IMPL(_TorchScriptTesting, CPU, m) {
+          m.impl("takes_foo", takes_foo);
+        }
+
+        # abstract impl in torchbind_impls.py:
+        @torch.library.impl_abstract("_TorchScriptTesting::takes_foo")
+        def foo_add_tensor(foo, z):
+            return foo.add_tensor(z)
+    """
+
+    def inner(fake_class):
+        global global_abstract_class_registry
```

Comment:
Don't think this is needed, `global` is  only necessary if you re-assign the variable.